src/main/scala/mesosphere/marathon/storage/repository/GcActor.scala

```diff
@@ -444,22 +453,25 @@ private[storage] trait CompactBehavior[K, C, S] extends StrictLogging { this: FS
           s"(${podVersionsToDelete.map { case (id, v) => id -> v.mkString("[", ", ", "]") }.mkString(", ")} as no roots refer to them" +
           " and they exceed max versions")
       }
-      val appFutures = appsToDelete.map(appRepository.delete)
-      val appVersionFutures = appVersionsToDelete.flatMap {
-        case (id, versions) =>
-          versions.map { version => appRepository.deleteVersion(id, version) }
-      }
-      val podFutures = podsToDelete.map(podRepository.delete)
-      val podVersionFutures = podVersionsToDelete.flatMap {
-        case (id, versions) =>
-          versions.map { version => podRepository.deleteVersion(id, version) }
-      }
-      val rootFutures = rootVersionsToDelete.map(groupRepository.deleteRootVersion)
-      await(Future.sequence(appFutures))
-      await(Future.sequence(appVersionFutures))
-      await(Future.sequence(podFutures))
-      await(Future.sequence(podVersionFutures))
-      await(Future.sequence(rootFutures))
+      val appFutures = Source(appsToDelete).mapAsync(ConcurrentCallLimit)(appRepository.delete).runWith(Sink.ignore)
+      val appVersionFutures = Source(appVersionsToDelete)
+        .flatMapConcat{
+          case (id, versions) =>
+            Source(versions).mapAsync(ConcurrentCallLimit) { version => appRepository.deleteVersion(id, version) }
+        }
+        .runWith(Sink.ignore)
+      val podFutures = Source(podsToDelete).mapAsync(ConcurrentCallLimit)(podRepository.delete).runWith(Sink.ignore)
+      val podVersionFutures = Source(podVersionsToDelete)
+        .flatMapConcat {
+          case (id, versions) =>
+            Source(versions).mapAsync(ConcurrentCallLimit) { version => podRepository.deleteVersion(id, version) }
+        }.runWith(Sink.ignore)
+      val rootFutures = Source(rootVersionsToDelete).mapAsync(ConcurrentCallLimit)(groupRepository.deleteRootVersion).runWith(Sink.ignore)
+      await(appFutures)
+      await(appVersionFutures)
+      await(podFutures)
+      await(podVersionFutures)
+      await(rootFutures)
```

Comment:
Maybe instead of kicking off all of these in parallel, we sequentialize them?

```scala
val appDeletes = Source(appsToDelete)
  .mapAsync(ConcurrentCallLimit)(appRepository.delete)

val appVersionDeletes = Source(appVersionsToDelete)
  .mapConcat { case (id, versions) => versions.map { id -> _ } }
  .mapAsync(ConcurrentCallLimit) { case (id, version) =>
    appRepository.deleteVersion(id, version)
  }

val podDelets = Source(podsToDelete)
  .mapAsync(ConcurrentCallLimit)(podRepository.delete)

val podVersionFutures = Source(podVersionsToDelete)
  .mapConcat { case (id, versions) => versions.map { id -> _ } }
  .mapAsync(ConcurrentCallLimit) { case (id, version) =>
    podRepository.deleteVersion(id, version)
  }

val rootDeletes = Source(rootVersionsToDelete)
  .mapAsync(ConcurrentCallLimit)(groupRepository.deleteRootVersion)


(appDeletes ++ appVersionDeletes ++ podDeletes ++ podVersionDeletes ++ rootDeletes)
  .runWith(Sink.ignore)
  .map { _ => CompactDone }
  .recover {
    case NonFatal(e) =>
      logger.error(s"While deleting unused objects ${Option(e.getMessage).getOrElse("")} encountered an error: ", e)
      CompactDone
  }
```

Reply:
Do you think it worth it?